diff --git a/vpp-manager/Makefile b/vpp-manager/Makefile
index bd7427b..005e9a2 100644
--- a/vpp-manager/Makefile
+++ b/vpp-manager/Makefile
@@ -59,6 +59,9 @@ clean-vpp:
 vpp: clone-vpp clean-vpp vpp-build-env
 	docker run --rm \
 		-e VPP_MGR_DIR=$(CURDIR) \
+		-e http_proxy=${http_proxy} \
+		-e https_proxy=${https_proxy} \
+		-e no_proxy=${no_proxy} \
 		-v $(CURDIR):$(CURDIR):delegated \
 		--user $$(id -u):$$(id -g) \
 		--network=host \
diff --git a/vpplink/binapi/vpp_clone_current.sh b/vpplink/binapi/vpp_clone_current.sh
index 65d9d5d..834a784 100755
--- a/vpplink/binapi/vpp_clone_current.sh
+++ b/vpplink/binapi/vpp_clone_current.sh
@@ -1,6 +1,8 @@
 #!/bin/bash
 set -e
 
+exit 0
+
 SCRIPTDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
 CACHE_DIR=$SCRIPTDIR/.cherries-cache
 
